File: lecture_011(RecursionBacktracking)/l_004_knighttour.cpp
#include<iostream>
#include<vector>
#include<string>
using namespace std;

void display(vector<vector<int>> &arr) {
    for(int i=0; i<arr.size(); i++) {
        for(int j=0; j<arr[0].size(); j++) {
            cout<<arr[i][j]<<" ";
        }
        cout<<endl;
    }
    cout<<endl;
}
void printknighttour(vector<vector<int>> &chess, int r, int c, int move) {
    if(r<0 || c<0 || r>=chess.size() || c>=chess.size() || chess[r][c]>0) {
        return;
    } 
    else if(move==chess.size()*chess.size()) {
        chess[r][c] = move;
        display(chess);
        chess[r][c] = 0;
        return;
    }
    chess[r][c] = move;
    printknighttour(chess,r-2,c+1,move+1);
    printknighttour(chess,r-1,c+2,move+1);
    printknighttour(chess,r+1,c+2,move+1);
    printknighttour(chess,r+2,c+1,move+1);
    printknighttour(chess,r+2,c-1,move+1);
    printknighttour(chess,r+1,c-2,move+1);
    printknighttour(chess,r-1,c-2,move+1);
    printknighttour(chess,r-2,c-1,move+1);
    chess[r][c] = 0;
}
int main() {
    int n, r, c;
    cin>>n>>r>>c;
    vector<vector<int>>chess(n,vector<int>(n,0));
    printknighttour(chess,r,c,1);

    return 0;
}

File: lecture_011(RecursionBacktracking)/l_004_knighttour.java
import java.util.Scanner;
public class l_004_knighttour {
    public static Scanner scn = new Scanner(System.in);
    public static void main(String[] args) {
        int n = scn.nextInt();
        int r = scn.nextInt();
        int c = scn.nextInt();
        int[][] chess = new int[n][n];
        printKnightTour(chess,r,c,1);
    }
    public static void display(int[][] arr) {
        for(int i=0; i<arr.length; i++) {
            for(int j=0; j<arr[0].length; j++) {
                System.out.print(arr[i][j]+" ");
            }
            System.out.println();
        }
        System.out.println();
    }
    public static void printKnightTour(int[][] chess, int r, int c, int move) {
        if(r<0 || c<0 || r>=chess.length || c>=chess.length || chess[r][c]>0) {
            return;
        } 
        else if(move==chess.length*chess.length) {
            chess[r][c] = move;
            display(chess);
            chess[r][c] = 0;
            return;
        }
        chess[r][c] = move;
        printKnightTour(chess,r-2,c+1,move+1);
        printKnightTour(chess,r-1,c+2,move+1);
        printKnightTour(chess,r+1,c+2,move+1);
        printKnightTour(chess,r+2,c+1,move+1);
        printKnightTour(chess,r+2,c-1,move+1);
        printKnightTour(chess,r+1,c-2,move+1);
        printKnightTour(chess,r-1,c-2,move+1);
        printKnightTour(chess,r-2,c-1,move+1);
        chess[r][c] = 0;
    }
}